pkg/operator/ceph/config/config.go

```diff
@@ -0,0 +1,195 @@
+package config
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/coreos/pkg/capnslog"
+	"github.com/go-ini/ini"
+	"github.com/rook/rook/pkg/clusterd"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var logger = capnslog.NewPackageLogger("github.com/rook/rook", "op-config")
+
+// DaemonType defines the type of a daemon. e.g., mon, mgr, osd, mds, rgw
+type DaemonType string
+
+const (
+	// MonType defines the mon DaemonType
+	MonType DaemonType = "mon"
+
+	// MgrType defines the mgr DaemonType
+	MgrType DaemonType = "mgr"
+
+	// OsdType defines the osd DaemonType
+	OsdType DaemonType = "osd"
+
+	// MdsType defines the mds DaemonType
+	MdsType DaemonType = "mds"
+
+	// RgwType defines the rgw DaemonType
+	RgwType DaemonType = "rgw"
+)
+
+// VarLibCephDir is simply "/var/lib/ceph". It is made overwriteable only for unit tests where it
+// may be needed to send data intended for /var/lib/ceph to a temporary test dir.
+var VarLibCephDir = "/var/lib/ceph"
+
+// EtcCephDir is simply "/etc/ceph". It is made overwriteable only for unit tests where it
+// may be needed to send data intended for /etc/ceph to a temporary test dir.
+var EtcCephDir = "/etc/ceph"
+
+// Config is a Ceph config struct containing zero or more Ceph config sections.
+type Config struct {
```

Comment:
Ceph config instead of being a strictly defined struct is now a map of maps where the first map is the config file section (e.g., global) and each section has a map of config key-value pairs. This is more flexible, but it is also potentially easier to break in code. Thoughts?

Reply:
What are all the scenarios where we anticipate this will be used? Is it just for the custom ceph.conf settings that users have today? And these settings would be read from the crd in the future? So these settings would be empty in the common case and only advanced users would set them, correct?

Reply:
As I understand your question, I anticipate this to be used to differing degrees in all scenarios. I wrote the config to look like this in part because it seems easier to take in override settings from the CRD (future work) and merge them with Rook's internal defaults.

I think this also enables the config Rook will set up for the Ceph cluster to be modified conditionally quite easily. As an example (pseudocode):
```go
c := NewConfig()

// establish default global settings
c.Section("global").
    Set("default setting", "blah").
    Set("other default setting", "blah")

UseMonCentralizedConfig = true

// override defaults for specific ceph versions
if ceph.Version == "luminous" {
    UseMonCentralizedConfig = false
    // apply some settings necessary only for luminous
    l := NewConfig()
    l.Section("global").Set("luminous setting", "blah")
    c.Merge(l)
} else if ceph.Version != "luminous" && ceph.Version != "mimic" {
    // apply some settings necessary only for luminous
    n:= NewConfig()
    n.Section("global").Set("nautilus and up setting", "blah")
    c.Merge(n)
}

// apply crd overrides
if crd.storage.overrides != nil && len(crd.storage.overrides) > 0 {
    c.Merge(crd.storage.overrides)
}

if ! UseMonCentralizedConfig {
    configFileStorage.UpdateOrCreate(c)
    return nil
}
// if using the centralized config, set these values with `ceph config ...` commands 
// later when at least one mon is active
return nil
```